### PR TITLE
Alerting: Failing to fetch rules to schedule from the database won't halt rule evaluation

### PR DESCRIPTION
**Pull request by gotjosh**
Grafana's alerting system should continue to run when failures at a database level are present. This change introduces a in-memory cache for the ready to schedule rules that we use when we're unable to fetch the newer versions from the database.

Partial solves #49855 but there's more to do.

I think we should do two more things (they can be separate PRs due to the urgency of Grafana 9) - if agreed, I'll move them to the main issue:

- [ ] Use the rules hash to determine whether we should update the cache or not
- [ ] Expose the hash as a metric (gauge in particular) to track changes to the ruleset over time.

### CODE REVIEW

File: pkg/services/ngalert/CHANGELOG.md [hunk 1]
```diff
@@ -51,6 +51,7 @@ Scopes must have an order to ensure consistency and ease of search, this helps u
   - `grafana_alerting_ticker_interval_seconds`
 - [ENHANCEMENT] Create folder 'General Alerting' when Grafana starts from the scratch #48866
 - [ENHANCEMENT] Rule changes authorization logic to use UID folder scope instead of ID scope #48970
+- [ENHANCEMENT] Failures at a database-level won't halt rule evaluation #XXX
```
Comment (gotjosh) [inline pkg/services/ngalert/CHANGELOG.md:54]:
```suggestion
- [ENHANCEMENT] Failures at a database-level won't halt rule evaluation #49857
```

File: pkg/services/ngalert/CHANGELOG.md [hunk 2]
```diff
@@ -51,6 +51,7 @@ Scopes must have an order to ensure consistency and ease of search, this helps u
   - `grafana_alerting_ticker_interval_seconds`
 - [ENHANCEMENT] Create folder 'General Alerting' when Grafana starts from the scratch #48866
 - [ENHANCEMENT] Rule changes authorization logic to use UID folder scope instead of ID scope #48970
+- [ENHANCEMENT] Failures at a database-level won't halt rule evaluation #49857
```
Comment (gotjosh) [inline pkg/services/ngalert/CHANGELOG.md:54]: Technically a BUGFIX but sadly the changes to this PR are buried among refactor commits and I couldn't pinpoint where exactly was the regression.

File: pkg/services/ngalert/schedule/fetcher.go [hunk 3]
```diff
@@ -16,10 +18,24 @@ func (sch *schedule) getAlertRules(ctx context.Context, disabledOrgs []int64) []
...
+
+// CacheRuleSet caches the current rule set in memory to avoid database failures halting alert evaluations.
+func (sch *schedule) CacheRuleSet(result []*models.SchedulableAlertRule) {
+	h := sha256.New()
+	for i := range result {
+		_, _ = h.Write([]byte(fmt.Sprintf("%v", result[i])))
+	}
+
+	sch.currentRuleSet = result
+	sch.currentRuleSetHash = fmt.Sprintf("%x", h.Sum(nil))
```
Comment (gotjosh) [inline pkg/services/ngalert/schedule/fetcher.go:40]: not used here, but hoping that the next PR can make use of this to avoid constant updates.

Comment (grobinson-grafana) [reply pkg/services/ngalert/schedule/fetcher.go:40]: I'm not sure `result` is sorted so the same set of alert rules will produce different hashes.

**Comment by gotjosh** [comment]
Superseded by [49874](https://github.com/grafana/grafana/pull/49874)

**Pull request closed without merging**